[strings] Fix split and rsplit implementations

bwasti opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23669 [strings] Fix split and rsplit implementations**

Differential Revision: [D16603705](https://our.internmc.facebook.com/intern/diff/D16603705)

Fixes #23616

bwasti committed:
[strings] Fix split and rsplit implementations

bwasti commented:
fixes #23616

driazati commented on test/test_jit_string.py:
```diff
@@ -263,9 +264,10 @@ def test_rpartition():
 
         def test_split():
-            # type: () -> Tuple[List[str], List[str], List[str], List[str], List[str], List[str], List[str], List[str], List[str]]
+            # type: () -> Tuple[List[str], List[str], List[str], List[str], List[str], List[str], List[str], List[str], List[str], List[str]]
```
Why not make this `List[List[str]]` so we don't have to keep editing it? Or even just delete the type comment since return types can be inferred

driazati commented on test/test_jit_string.py:
```diff
@@ -1,5 +1,6 @@
 from test_jit import JitTestCase
-
+import sys
+sys.setrecursionlimit(10000)
```
Why is this needed?

bwasti replied:
it won't pass the tests on my machine without it -- the default seems to conservative

suo replied:
what is causing the big recursion? Is it not suspicious?

driazati commented on torch/csrc/jit/register_string_ops.cpp:
```diff
@@ -547,48 +547,94 @@ auto reg_str_ops_2 = torch::RegisterOperators()
+    // "__DO_NOT_USE" as the default separator for split functions.
+    // This denotes splitting by any whitespace, or default behavior when
+    // the separator is not specified
+    .op("aten::split(str self, str separator='__DO_NOT_USE', int max=-1) -> str[]",
```
Can `separator` be type `str?` instead?

bwasti replied:
ooh, thanks! didn't realize this existed -- cleans up a lot

driazati commented on torch/csrc/jit/register_string_ops.cpp:
```diff
@@ -547,48 +547,94 @@ auto reg_str_ops_2 = torch::RegisterOperators()
+              if (max >= 0 && count > max) {
+                break;
+              } else {
+                splits.emplace_back(string.substr(prev_pos, pos - prev_pos));
```
For a string like `a\t\tb` won't this lead to `['a', '', 'b']`? Python treats all contiguous whitespace as 1 separator

bwasti replied:
ah yea, nice catch! this is true if the separator isn't specified (and not true if it is)

driazati commented on torch/csrc/jit/register_string_ops.cpp:
```diff
@@ -547,48 +547,94 @@ auto reg_str_ops_2 = torch::RegisterOperators()
+            separators = { "\n", "\t", " " };
+          }
+          do {
+            auto min_pos = std::string::npos;
```
there's a lot of code copy paste between here and rsplit, could they be combined?

bwasti committed:
Update on "[strings] Fix split and rsplit implementations"

[strings] Fix split and rsplit implementations

gh-metadata: pytorch pytorch 23669 gh/bwasti/30/head

suo commented on test/test_jit_string.py:
```diff
@@ -1,5 +1,6 @@
 from test_jit import JitTestCase
```
these tests are not run in oss CI. You need to whitelist them in `run_tests.py`.

bwasti committed:
Update on "[strings] Fix split and rsplit implementations"

[strings] Fix split and rsplit implementations

gh-metadata: pytorch pytorch 23669 gh/bwasti/30/head

bwasti committed:
Update on "[strings] Fix split and rsplit implementations"

[strings] Fix split and rsplit implementations

gh-metadata: pytorch pytorch 23669 gh/bwasti/30/head

suo commented on test/test_jit_string.py:
```diff
@@ -278,18 +279,20 @@ def test_split():
+                " j*j j*j j".rsplit("j*", 10),
             )
+
         self.checkScript(test_rsplit, ())
```
can you also add an `if __name__ == __main__`  thing at the bottom so we can run this test in isolation?

suo commented:
@bwasti any update on this?

bwasti committed:
Update on "[strings] Fix split and rsplit implementations"

[strings] Fix split and rsplit implementations

gh-metadata: pytorch pytorch 23669 gh/bwasti/30/head

suo reviewed (approved):
lgtm except there are test failures looks like

bwasti commented:
this PR is now covered by #25422
